inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -42,572 +42,547 @@
+}
 
-        if (std::is_same<T, float>::value && chs == 4) {
-            avx::splitRow_32FC4(reinterpret_cast<const float*>(in),
-                                reinterpret_cast<float*>(outs[0]),
-                                reinterpret_cast<float*>(outs[1]),
-                                reinterpret_cast<float*>(outs[2]),
-                                reinterpret_cast<float*>(outs[3]),
-                                length);
-            return;
-        }
-    }
-#endif  // HAVE_AVX2
+namespace {
+using merge_supported_types = typelist<uint8_t, int8_t, uint16_t, int16_t, int32_t, float, fp_16_t>;
 
-#ifdef HAVE_SSE
-    if (with_cpu_x86_sse42()) {
-        if (std::is_same<T, uint8_t>::value && chs == 2) {
-            splitRow_8UC2(in, outs[0], outs[1], length);
-            return;
+template<typename T, int chs>
+void mergeRowImpl(scalar_tag, const std::array<const T*, chs>& ins, T* out, const int length) {
+    for (int x = 0; x < length; ++x) {
+        for (int c = 0; c < chs; ++c) {
+            out[chs * x + c] = ins[c][x];
         }
+    }
+}
 
-        if (std::is_same<T, uint8_t>::value && chs == 3) {
-            splitRow_8UC3(in, outs[0], outs[1], outs[2], length);
-            return;
-        }
+template<typename isa_tag_t, int chs>
+struct typed_merge_row {
+    using p_f = void (*)(const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length);
+
+    template <typename type, typename tag = isa_tag_t>
+    typename std::enable_if<std::is_same<tag, scalar_tag>::value ||
+            (!std::is_same<tag, scalar_tag>::value && !std::is_same<type, uint8_t>::value &&
+            !std::is_same<type, float>::value), p_f>
+            ::type operator()(type_to_type<type> ) {
+            return [](const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length) {
+            const auto inT = reinterpret_cast<const std::array<const type*, chs>&>(ins);
+            auto outT = reinterpret_cast<type*>(out);
+            scalar_tag t;
+            mergeRowImpl<type, chs>(t, inT, outT, length);
+        };
+    }
+#if MANUAL_SIMD
+    template<typename type, typename tag = isa_tag_t>
+    typename
+        std::enable_if<!std::is_same<tag, scalar_tag>::value&& std::is_same<type, uint8_t>::value, p_f>
+        ::type operator()(type_to_type<uint8_t>) {
+        return [](const std::array<const uint8_t*, chs>& ins, uint8_t* out, const int length) {
+            isa_tag_t t;
+            mergeRowImpl<isa_tag_t, uint8_t, chs>(t, ins, out, length);
+        };
+    }
 
-        if (std::is_same<T, uint8_t>::value && chs == 4) {
-            splitRow_8UC4(in, outs[0], outs[1], outs[2], outs[3], length);
-            return;
-        }
+    template<typename type, typename tag = isa_tag_t>
+    typename
+        std::enable_if<!std::is_same<tag, scalar_tag>::value&& std::is_same<type, float>::value, p_f>
+        ::type operator()(type_to_type<float>) {
```

Comment:
looks like the first template param `type` is useless here and/or cannot  be deducted,
because you use `overload`  version which receives explicit param 

operator() (type_to_type<**float**>)

could you change it on

operator() (type_to_type<**type**>)